File: content/LLD/Reference-Code/3. Memento Pattern/history.py
from text_memento import TextMemento
from typing import List


class History:
    def __init__(self):
        self.__history: List[TextMemento] = []

    def save_state(self, tm: TextMemento):
        self.__history.append(tm)

    def undo(self) -> TextMemento:
        if len(self.__history) > 0:
            self.__history.pop()
            if len(self.__history) == 0:
                return TextMemento("")
            return self.__history[-1]
        else:
            return TextMemento("")

    def get_history(self):
        for i in range(len(self.__history)):
            print(f"{i} = {self.__history[i].get_saved_text()}")

File: content/LLD/Reference-Code/3. Memento Pattern/History.java
import java.util.ArrayList;
import java.util.List;

public class History {
    private final List<TextMemento> history = new ArrayList<>();

    public void saveState(TextMemento tm) {
        this.history.add(tm);
    }

    public TextMemento undo() {
        if (!this.history.isEmpty()) {
            this.history.remove(this.history.size() - 1);
            if (this.history.isEmpty()) {
                return new TextMemento("");
            }
            return this.history.get(this.history.size() - 1);
        } else {
            return new TextMemento("");
        }
    }

    public void getHistory() {
        for (int i = 0; i < this.history.size(); i++) {
            System.out.println(i + " = " + this.history.get(i).getSavedText());
        }
    }
}
